lib/Dialect/FIRRTL/Transforms/DropConst.cpp

```diff
@@ -0,0 +1,128 @@
+/// Return this type with a 'const' modifiers dropped
+static FIRRTLBaseType getAllConstDroppedType(FIRRTLBaseType type) {
+  if (!type.containsConst())
+    return type;
+  return TypeSwitch<FIRRTLBaseType, FIRRTLBaseType>(type)
+      .Case<ClockType, ResetType, AsyncResetType, AnalogType, SIntType,
+            UIntType>([](auto type) { return type.getConstType(false); })
+      .Case<BundleType>([](BundleType type) {
+        SmallVector<BundleType::BundleElement> constDroppedElements(
+            llvm::map_range(
+                type.getElements(), [](BundleType::BundleElement element) {
+                  element.type = getAllConstDroppedType(element.type);
+                  return element;
+                }));
+        return BundleType::get(type.getContext(), constDroppedElements, false);
+      })
+      .Case<FVectorType>([](FVectorType type) {
+        return FVectorType::get(getAllConstDroppedType(type.getElementType()),
+                                type.getNumElements(), false);
+      })
+      .Case<FEnumType>([](FEnumType type) {
+        SmallVector<FEnumType::EnumElement> constDroppedElements(
+            llvm::map_range(
+                type.getElements(), [](FEnumType::EnumElement element) {
+                  element.type = getAllConstDroppedType(element.type);
+                  return element;
+                }));
+        return FEnumType::get(type.getContext(), constDroppedElements, false);
+        ;
+      })
+      .Default([](Type) {
+        llvm_unreachable("unknown FIRRTL type");
+        return FIRRTLBaseType();
+      });
+}
+// NOLINTEND(misc-no-recursion)
+
+/// Returns null type if no conversion is needed.
+static FIRRTLBaseType convertType(FIRRTLBaseType type) {
+  auto nonConstType = getAllConstDroppedType(type);
+  return nonConstType != type ? nonConstType : FIRRTLBaseType{};
+}
+
+/// Returns null type if no conversion is needed.
+static Type convertType(Type type) {
+  if (auto base = type.dyn_cast<FIRRTLBaseType>()) {
+    return convertType(base);
+  }
+
+  if (auto refType = type.dyn_cast<RefType>()) {
+    if (auto converted = convertType(refType.getType()))
+      return RefType::get(converted, refType.getForceable());
+  }
+
+  return {};
+}
+
+namespace {
+class DropConstPass : public DropConstBase<DropConstPass> {
+  void runOnOperation() override {
+    auto module = getOperation();
+    auto fmodule = dyn_cast<FModuleOp>(*module);
+
+    // Convert the module body if present
+    if (fmodule) {
+      fmodule->walk([](Operation *op) {
+        if (auto constCastOp = dyn_cast<ConstCastOp>(op)) {
+          // Remove any `ConstCastOp`, replacing results with inputs
+          constCastOp.getResult().replaceAllUsesWith(constCastOp.getInput());
+          constCastOp->erase();
+          return;
+        }
+
+        for (auto result : op->getResults()) {
+          if (auto convertedType = convertType(result.getType()))
+            result.setType(convertedType);
+        }
```

Comment:
The MatchOp has block arguments, so you should probably update those if the operation has any.  I think this would also update the top level-modules block arguments as well.

Reply:
The match op is lowered before this pass, but I don't think it hurts to generically check regions. Updated.